fractal_tasks_core/lib_ngff.py

```diff
@@ -0,0 +1,412 @@
+
+    @property
+    def pixel_sizes_zyx(self) -> list[list[float]]:
+        """
+        Pixel sizes extracted from scale transformations of datasets.
+
+        Raises:
+            ValueError:
+                If pixel sizes are below a given threshold (1e-9).
+        """
+        x_index = self.axes_names.index("x")
+        y_index = self.axes_names.index("y")
+        try:
+            z_index = self.axes_names.index("z")
+        except ValueError:
+            z_index = None
+            logging.warning(
+                f"Z axis is not present (axes: {self.axes_names}), and Z pixel"
+                " size is set to 1. This may work, by accident, but it is "
+                "not fully supported."
+            )
+        _pixel_sizes_zyx = []
+        for level in range(self.num_levels):
+            scale = self.datasets[level].scale_transformation.scale
+            pixel_size_x = scale[x_index]
+            pixel_size_y = scale[y_index]
+            if z_index is not None:
+                pixel_size_z = scale[z_index]
+            else:
+                pixel_size_z = 1.0
+            _pixel_sizes_zyx.append([pixel_size_z, pixel_size_y, pixel_size_x])
+            if min(_pixel_sizes_zyx[-1]) < 1e-9:
+                raise ValueError(
+                    f"Pixel sizes at level {level} are too small: "
+                    f"{_pixel_sizes_zyx[-1]}"
+                )
+
+        return _pixel_sizes_zyx
+
+    def get_pixel_sizes_zyx(self, *, level: int = 0) -> list[float]:
+        return self.pixel_sizes_zyx[level]
+
+    @property
+    def coarsening_xy(self) -> int:
+        """
+        Linear coarsening factor in the YX plane.
+
+        We only support coarsening factors that are homogeneous (both in the
+        X/Y directions and across pyramid levels).
+
+        Raises:
+            NotImplementedError:
+                If coarsening ratios are not homogeneous.
+        """
+        current_ratio = None
+        for ind in range(1, self.num_levels):
+            ratio_x = round(
+                self.pixel_sizes_zyx[ind][2] / self.pixel_sizes_zyx[ind - 1][2]
+            )
+            ratio_y = round(
+                self.pixel_sizes_zyx[ind][1] / self.pixel_sizes_zyx[ind - 1][1]
+            )
+            if ratio_x != ratio_y:
+                raise NotImplementedError(
+                    "Inhomogeneous coarsening in X/Y directions "
+                    "is not supported.\n"
+                    f"ZYX pixel sizes:\n {self.pixel_sizes_zyx}"
+                )
+            if current_ratio is None:
+                current_ratio = ratio_x
+            else:
+                if current_ratio != ratio_x:
+                    raise NotImplementedError(
+                        "Inhomogeneous coarsening across levels "
+                        "is not supported.\n"
+                        f"ZYX pixel sizes:\n {self.pixel_sizes_zyx}"
+                    )
+
+        return current_ratio
+
+
+class Image(BaseModel):
```

Comment:
This image class is a bit confusing to me. It does not contain the actual image data, but the metadata that goes into the well .zattrs for a single image. Is it our naming convention or an NGFF naming convention? Let's at least make the docstring clearer to specify this

Reply:
The specs have an `images` attribute of a well (https://github.com/ome/ngff/blob/main/0.4/schemas/well.schema), and then the most natural way to name the model is indeed `Image`. But I agree: this is confusing and may overlap with `NgffImageMeta`. I'm open to any suggestion on a different name (e.g `WellImage`, `ImageInWell`, ..).

More in general, we could mark all "internal" model names with an underscore (`_Image`, `_Multiscale`, `_Axis`, ...), since we only expect tasks to import/use the `Ngff...Meta` ones (i.e. only `NgffImageMeta` and `NgffWellmeta`, for the moment) while the other are meant as models for the nested properties.

The docstring says "Model for an element of `Well.images`.", but I am now also adding an additional note.

Reply:
Hmm, I'd be a bit hesitant to make too many of the models internal. The idea would be that these Pydantic models are how someone should interact with the spec, right?
How about `ImageInWellMeta`?

Reply:
Docstring update: b7107e265a2b5c398b8de71df48ce06c38fbd5d1

Reply:
Call output: name will be `ImageInWell`

Reply:
Tasks use the `NgffImageMeta` model